client/rpc/server/server.go

```diff
@@ -0,0 +1,975 @@
+		}
+
+		// Read and respond to the request.
+		s.jsonRPCRead(w, r, isAdmin)
+	})
+
+	// Websocket endpoint.
+	rpcServeMux.HandleFunc("/ws", func(w http.ResponseWriter, r *http.Request) {
+		authenticated, isAdmin, err := s.checkAuth(r, false)
+		if err != nil {
+			jsonAuthFail(w)
+			return
+		}
+
+		// Attempt to upgrade the connection to a websocket connection
+		// using the default size for read/write buffers.
+		ws, err := websocket.Upgrade(w, r, nil, 0, 0)
+		if err != nil {
+			if _, ok := err.(websocket.HandshakeError); !ok {
+				log.Errorf("Unexpected websocket error: %v",
+					err)
+			}
+			http.Error(w, "400 Bad Request.", http.StatusBadRequest)
+			return
+		}
+		ws.SetPingHandler(func(payload string) error {
+			err := ws.WriteControl(websocket.PongMessage, []byte(payload),
+				time.Now().Add(time.Second))
+			log.Debugf("ping received: len %d", len(payload))
+			log.Tracef("ping payload: %s", payload)
+			if err != nil {
+				log.Errorf("Failed to send pong: %v", err)
+				return err
+			}
+			return nil
+		})
+		ws.SetPongHandler(func(payload string) error {
+			log.Debugf("pong received: len %d", len(payload))
+			log.Tracef("pong payload: %s", payload)
+			return nil
+		})
+		s.WebsocketHandler(ws, r.RemoteAddr, authenticated, isAdmin)
+	})
+
+	for _, listener := range s.cfg.Listeners {
+		s.wg.Add(1)
+		go func(listener net.Listener) {
+			log.Infof("RPC server listening on %s",
+				listener.Addr())
+			httpServer.Serve(listener)
+			log.Tracef("RPC listener done for %s",
+				listener.Addr())
+			s.wg.Done()
+		}(listener)
+	}
+
+	s.ntfnMgr.Start()
+}
+
+// genCertPair generates a key/cert pair to the paths provided.
+func genCertPair(certFile, keyFile string, altDNSNames []string) error {
+	log.Infof("Generating TLS certificates...")
+
+	org := "dex autogenerated cert"
+	validUntil := time.Now().Add(10 * 365 * 24 * time.Hour)
+	cert, key, err := certgen.NewTLSCertPair(elliptic.P521(), org,
+		validUntil, altDNSNames)
+	if err != nil {
+		return err
+	}
+
+	// Write cert and key files.
+	if err = ioutil.WriteFile(certFile, cert, 0644); err != nil {
+		return err
+	}
+	if err = ioutil.WriteFile(keyFile, key, 0600); err != nil {
+		os.Remove(certFile)
+		return err
+	}
+
+	log.Infof("Done generating TLS certificates")
+	return nil
+}
+
+// Config is a descriptor containing the RPC server configuration.
+type Config struct {
+	// Listeners defines a slice of listeners for which the RPC server will
+	// take ownership of and accept connections.  Since the RPC server takes
+	// ownership of these listeners, they will be closed when the RPC server
+	// is stopped.
+	Listeners            []net.Listener
+	RPCListeners         []string
+	RPCUser              string
+	RPCPass              string
+	RPCKey               string
+	RPCCert              string
+	AltDNSNames          []string
+	RPCMaxClients        int
+	RPCMaxConcurrentReqs int
+	RPCMaxWebsockets     int
+	RPCLimitUser         string
+	RPCLimitPass         string
```

Comment:
Configuration should be simplified a lot. Here's the constructor for a `webserver.WebServer`

`func New(core clientCore, addr string, logger slog.Logger, reloadHTML bool)`

The RPC server should be about as easy. We can make things more complicated later if a need arises.